src/fges.jl

```diff
@@ -0,0 +1,536 @@
+        # If the score did not improve...
+        if newStep.Δscore ≤ 0
+            #...stop searching
+            break
+        end
+        
+        # Use the insert or delete operator update the graph
+        operator(g, newStep)
+        
+        # Covert the PDAG to a complete PDAG
+        # Undirect all edges unless they participate in a v-structure
+        graphVStructure!(g)
+        # Apply the 4 Meek rules to orient some edges in the graph
+        meekRules!(g)
+
+        # Reset the score
+        newStep.Δscore = 0
+
+        
+    end
+
+    return nothing
+end
+
+
+function findNextEquivClass!(newStep, data, g, findBestOperation::Function, verbose)
+
+    # Parallelization with synced threads (to avoid race conditions)
+    @sync begin
+        # Loop through all possible node combinations
+        for (x,y) in allpairs(vertices(g)) 
+            # Spawn a new thread
+            Threads.@spawn begin
+                # Check if there is an edge between two vertices
+                hasEdge = isadjacent(g,x,y)
+
+                # Skip over adjacent edges if we're trying to insert and vice versa
+                if findBestOperation == findBestInsert ? !hasEdge : hasEdge
+                    #For this pair of nodes, the following function will:
+                        #(1) check if a valid operator exists
+                        #(2) score all valid operators and return the one with the highest score
+                        #findBestOperation is either "findBestInsert" or "findBestDelete"
+                    (bestSubset, bestScore) = findBestOperation(newStep, data, g, x, y)
+
+                    # If the best valid operator was better than any previously found...
+                    if bestScore > newStep.Δscore 
+                        #...update newStep
+                        newStep.edge = Edge(x,y)
+                        newStep.subset = bestSubset
+                        newStep.Δscore = bestScore
+
+                        verbose && println(newStep)
+                    end
+                end
+
+            end
+        end
+    end
+end
+
+
+function findBestInsert(step, dataParsed, g, x, y)
+    # Calculate two (possibly empty) sets of nodes
+    # NAxy: any nodes that are undirected neighbors of y and connected to x by any edge
+    # Txy: any subset of the undirected neighbors of y not connected to x
+    Tyx = calcT(g, y, x)
+    NAyx = calcNAyx(g, y, x)
+
+
+    # Create two containers to hold the best found score and best subset of Tyx
+    bestScore = zero(step.Δscore)
+    bestT = Vector{Int}()
+
+    # Keep a list of invalid sets
+    invalid = Vector{Vector{Int}}()
+    
+    # Loop through all possible subsets of Tyx
+    for T in powerset(Tyx)
+        if checkSupersets(T,invalid)
+            NAyxT = NAyx ∪ T
+            if isclique(g, NAyxT) && isblocked(g, y, x, NAyxT)
+
+                # Score the valid Insert
+                PAy = parents(g,y)
+                PAy⁺ = PAy ∪ x
```

Comment:
I like the use of ∪, it's very readable, we could use this in other places of the package as well

Reply:
Coming out of the woodwork again. I also like this notation because it is easy to read, but it does create more allocations. I've tried using `Sets` but you can't (easily?) uses them for indexing arrays.